File: NeuralNet/NeuralNetTypes/BlockType/Blocks/SimpleNeuronBlock.cs
using System;
using System.Threading.Tasks;

namespace NeuralNet {
	[Serializable]
	public sealed class SimpleNeuronBlock : BaseNeuralBlock {
		public SimpleNeuronBlock(int size, BaseNeuralBlock parent, IActivationFunction activationFunction)
			: base(size, new[] {parent}, activationFunction) {}

		public SimpleNeuronBlock(int size, int parentSize, IActivationFunction activationFunction)
			: base(size, parentSize, activationFunction) {}

		public override void Calculate() {
			var parent = Parents[0];
			var parentState = parent.GetState();
			var parentSize = parent.Size;
			var weightsForParent = Weights[0];
			var neuronsCount = State.Length;

			Parallel.For(0, neuronsCount, neuronNum => {
				var sum = 0.0f;
				for (var i = 0; i < parentState.Length; i++) {
					sum += parentState[i]
					       *weightsForParent[neuronNum*parentSize + i];
				}
				Net[neuronNum] = sum + Bias[neuronNum];
				State[neuronNum] = ActivationFunction.Calculate(sum);
			});
		}

		public override void Calculate(float[] input) {
			var firstParentBlockWeights = Weights[0];
			var inputSize = input.Length;
			var neuronsCount = State.Length;

			Parallel.For(0, neuronsCount, neuronNum => {
				var sum = 0.0f;
				for (var i = 0; i < inputSize; i++) {
					sum += input[i]*firstParentBlockWeights[neuronNum*inputSize + i];
				}
				Net[neuronNum] = sum + Bias[neuronNum];
				State[neuronNum] = ActivationFunction.Calculate(sum);
			});
		}
	}
}

File: NeuralNetNative/SimpleNeuronBlock.cpp
#define NEURALNETNATIVEAPI
#include "SimpleNeuronBlock.h"
#include <tbb\tbb.h>
#include <tbb\task_scheduler_init.h>
#include <tbb\parallel_for.h>
#include <tbb\blocked_range.h>
#include "GrainSizeForParallel.h"

using namespace tbb;

namespace NeuralNetNative {
	SimpleNeuronBlock::SimpleNeuronBlock(int size, BaseNeuralBlock *parent, ActivationFunction *function) : BaseNeuralBlock(size, parent, function) {
	}

	SimpleNeuronBlock::SimpleNeuronBlock(int size, int parentSize, ActivationFunction *function) : BaseNeuralBlock(size, parentSize, function) {
	}

	void SimpleNeuronBlock::Calculate(void) {
		float *parentState = Parent->GetState();
		int parentSize = Parent->GetSize();
		
		parallel_for(blocked_range<size_t>(0, Size, SimpleNeuronBlockGrainSize),
		[=](const blocked_range<size_t>& r)
		{
			for (int neuronNum = r.begin(); neuronNum < r.end(); neuronNum++) {
				float sum = 0.0f;
				#pragma simd
				for (int i = 0; i < parentSize; i++) {
					sum += parentState[i]*Weights[neuronNum*parentSize + i];
				}
				Net[neuronNum] = sum + Bias[neuronNum];
				State[neuronNum] = Function->Calculate(sum);
			}
		});
	}

	void SimpleNeuronBlock::Calculate(const float *input) {
		parallel_for(blocked_range<size_t>(0, Size, SimpleNeuronBlockGrainSize),
		[=](const blocked_range<size_t>& r)
		{
			for (int neuronNum = r.begin(); neuronNum < r.end(); neuronNum++) {
				float sum = 0.0f;
				#pragma simd
				for (int i = 0; i < PreviousSize; i++) {
					sum += input[i]*Weights[neuronNum*PreviousSize + i];
				}
				Net[neuronNum] = sum + Bias[neuronNum];
				State[neuronNum] = Function->Calculate(sum);
			}
		});
	}
}